File: xwm.osgi-starter/src/main/java/xwm/xwm/osgistarter/BundleLoader.java
package xwm.xwm.osgistarter;

import org.eclipse.core.runtime.adaptor.EclipseStarter;
import org.osgi.framework.Bundle;
import org.osgi.framework.BundleContext;
import org.osgi.framework.BundleException;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class BundleLoader {

   static BundleLoader instance = new BundleLoader();

   private final BundleContext bc;
   //   private final Framework fwk;

   private BundleLoader() {

      try {
         Map<String, String> config = new HashMap<String, String>();
         config.put("org.osgi.framework.bootdelegation", "*");
         config.put("osgi.parentClassloader", "fwk");
         config.put("osgi.contextClassLoaderParent", "fwk");

         config.put("osgi.bundles", "org.eclipse.equinox.common@2:start,org.eclipse.update" +
            ".configurator@3:start,org.eclipse.core.jobs@4:start,org.eclipse.core.runtime@start," +
            "org.eclipse.equinox.registy@start");
         config.put("osgi.clean", "true");
         config.put("eclipse.noRegistryCache", "true");

         EclipseStarter.setInitialProperties(config);
         bc = EclipseStarter.startup(new String[0], null);

         // TODO get Eclipse Platform for extension points

      } catch (Exception e) {
         throw new RuntimeException(e);
      }
   }

   public static Bundle installBundle(String file) throws BundleException {
      if (!file.startsWith("file:")) {
         file = "file:" + file;
      }
      Bundle bundle = instance.bc.installBundle(file);

      return bundle;
   }

   public static boolean uninstallBundle(long id) throws BundleException {
      Optional<Bundle> first = Stream.of(instance.bc.getBundles())
         .filter(b -> b.getBundleId() == id)
         .findFirst();

      if (first.isPresent()) {
         first.get()
            .uninstall();
         return true;
      } else {
         return false;
      }
   }

   public static List<Bundle> getBundles() {
      return Stream.of(Optional.ofNullable(instance.bc.getBundles())
            .orElse(new Bundle[0]))
         .collect(Collectors.toList());
   }

}

File: xwm.osgi-starter/src/main/java/xwm/xwm/osgistarter/BundleLoader.kt
package xwm.xwm.osgistarter

import org.eclipse.core.runtime.Platform
import org.eclipse.core.runtime.adaptor.EclipseStarter
import org.osgi.framework.Bundle
import org.osgi.framework.BundleContext
import org.osgi.framework.BundleException
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

class BundleLoader private constructor() {
   private var bc: BundleContext? = null

   //   private final Framework fwk;
   init {
      try {
         val config: MutableMap<String, String> = HashMap()
         config["org.osgi.framework.bootdelegation"] = "*"
         config["osgi.parentClassloader"] = "fwk"
         config["osgi.contextClassLoaderParent"] = "fwk"
         config["osgi.bundles"] = "org.eclipse.equinox.common@2:start,org.eclipse.update" +
            ".configurator@3:start,org.eclipse.core.jobs@4:start,org.eclipse.core.runtime@start," + "org.eclipse.equinox.registy@start"
         config["osgi.clean"] = "true"
         config["eclipse.noRegistryCache"] = "true"
         EclipseStarter.setInitialProperties(config)
         bc = EclipseStarter.startup(arrayOfNulls(0), null)

         val extensionPoints = Platform.getExtensionRegistry().extensionPoints
         println(extensionPoints)

         // TODO get Eclipse Platform for extension points
      } catch (e: Exception) {
         throw RuntimeException(e)
      }
   }

   companion object {
      private var instance = BundleLoader()

      @JvmStatic
      @Throws(BundleException::class)
      fun installBundle(file: String): Bundle {
         var file = file
         if (!file.startsWith("file:")) {
            file = "file:$file"
         }
         return instance.bc!!.installBundle(file)
      }

      @JvmStatic
      @Throws(BundleException::class)
      fun uninstallBundle(id: Long): Boolean {
         val first = Stream.of(*instance.bc!!.bundles)
            .filter { b: Bundle -> b.bundleId == id }
            .findFirst()
         return if (first.isPresent) {
            first.get()
               .uninstall()
            true
         } else {
            false
         }
      }

      @JvmStatic
      val bundles: List<Bundle>
         get() = Stream.of(
            *Optional.ofNullable(
               instance.bc!!.bundles
            )
               .orElse(arrayOfNulls(0))
         )
            .collect(Collectors.toList())
   }
}